include/picongpu/particles/Particles.kernel

```diff
@@ -204,65 +204,146 @@ struct KernelDeriveParticles
+            )
+            {
+                frame = pb.getLastFrame( superCellIdx );
+            }
+        );
 
         __syncthreads();
-        if (!frame.isValid())
-            return; //end kernel if we have no frames
 
-        /* BUGFIX to issue #538
-         * volatile prohibits that the compiler creates wrong code*/
-        volatile bool isParticle = (*frame)[linearThreadIdx][multiMask_];
+        // end kernel if we have no frames
+        if( !frame.isValid( ) )
+            return;
+
+        using ParticleDomCfg = IdxConfig<
+            frameSize,
+            numWorkers
+        >;
 
-        /* offset of the superCell (in cells, without any guards) to the origin of the local domain */
-        const DataSpace<simDim> localSuperCellOffset =
-            superCellIdx - mapper.getGuardingSuperCells();
+        // marker if a particle slot within a frame holds a valid particle
+        memory::CtxArray<
+            bool,
+            ParticleDomCfg
+        >
+        isParticleCtx(
+            workerIdx,
+            [&](
+                uint32_t const linearIdx,
+                uint32_t const
+            )
+            {
+                return frame[ linearIdx ][ multiMask_ ];
+            }
+        );
 
-        auto accManipulator = particleFunctor(
+        // offset of the superCell (in cells, without any guards) to the origin of the local domain
+        DataSpace< simDim > const localSuperCellOffset =
+            superCellIdx - mapper.getGuardingSuperCells( );
+
+        auto accManipulator = manipulateFunctor(
             localSuperCellOffset,
-            mappings::threads::WorkerCfg< pmacc::math::CT::volume<
-                SuperCellSize
-            >::type::value >( linearThreadIdx )
+            WorkerCfg< T_numWorkers >{ workerIdx }
         );
-        __syncthreads();
 
-        while (frame.isValid())
+        __syncthreads( );
+
+        while( frame.isValid( ) )
         {
-            auto particle = frame[linearThreadIdx];
-            if( !isParticle )
-                particle.setHandleInvalid( );
-            accManipulator( particle );
+            // loop over all particles in the frame
+            ForEachIdx< ParticleDomCfg >{ workerIdx }(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    auto particle = frame[ linearIdx ];
+                    bool const isPar = isParticleCtx[ idx ];
 
-            __syncthreads();
-            if (linearThreadIdx == 0)
-            {
-                frame = pb.getPreviousFrame(frame);
-            }
-            isParticle = true;
-            __syncthreads();
+                    if( !isPar )
+                        particle.setHandleInvalid( );
+
+                    // call manipulator even if the particle is not valid
+                    accManipulator( particle );
+
+                    // assume that all other frames has no gaps
```

Comment:
```C++
/* only the last frame is allowed to be non-full: all following
 * frames' particles will be valid, since we iterate the list of
 * frames from back to front
 */
```